File: examples/c99/two_runtimes.c
/*
 * two_runtimes.c — running two jse runtimes side by side in one process.
 *
 * Where main.c drives a single runtime, this opens two and shows what they do
 * and do not share:
 *
 *   1. globals    — a global set in A is invisible in B
 *   2. objects    — each runtime builds its own, with its own shapes
 *   3. strings    — the same literal interns separately in each
 *   4. handles    — a jse_value belongs to one runtime, and the engine does
 *                   not reliably catch you for using it with the other
 *   5. lifetime   — closing A leaves B untouched
 *
 * Build and run with `make run-two-runtimes` (see README.md).
 */

#include "jse_util.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Evaluate `src` in `rt`, print `label = result`, and report any failure. */
static int show(jse_runtime rt, const char *label, const char *src)
{
    char *text = jseu_eval_to_string(rt, src);

    if (text == NULL) {
        printf("%-30s ! %s\n", label, jse_last_error(rt));
        return 0;
    }
    printf("%-30s = %s\n", label, text);
    free(text);
    return 1;
}

int main(void)
{
    jse_runtime a = NULL;
    jse_runtime b = NULL;
    jse_value   from_a = JSE_INVALID_VALUE;
    double      n = 0.0;
    int         status;

    /*
     * ------------------------------------------------------ 1. two runtimes
     *
     * Nothing is process-global, so jse_open succeeds as many times as you ask
     * it to. Each runtime owns its own globals, objects, shapes and interned
     * strings, and they are independent for the whole of their lifetimes.
     *
     * Each must still be driven from one thread at a time: the engine has no
     * locking. Two threads each driving their OWN runtime share nothing and
     * are fine; two threads inside one runtime are not, and nothing stops you.
     */
    if (jse_open(&a) != JSE_OK || jse_open(&b) != JSE_OK) {
        fprintf(stderr, "jse_open failed\n");
        jse_close(a);
        jse_close(b);
        return EXIT_FAILURE;
    }
    printf("jse version %s\n\n", jse_version());

    /*
     * ------------------------------------------------------- 2. globals
     *
     * The same name, a different value in each, and neither leaks into the
     * other's global object.
     */
    printf("independent globals:\n");
    jseu_eval_cstr(a, "var tag = 'A'; var n = 111;", NULL);
    jseu_eval_cstr(b, "var tag = 'B'; var n = 222;", NULL);
    show(a, "  A.tag / A.n", "tag + '/' + n");
    show(b, "  B.tag / B.n", "tag + '/' + n");
    jseu_eval_cstr(b, "var onlyB = 1;", NULL);
    show(b, "  B.onlyB", "typeof globalThis.onlyB");
    show(a, "  A.onlyB", "typeof globalThis.onlyB");

    /*
     * -------------------------------------------------------- 3. objects
     *
     * Both build an object through the same property sequence, which drives
     * the same shape transitions in each. Separate shape tables mean the two
     * do not interfere; each reads back exactly what it wrote.
     */
    printf("\nindependent objects and shapes:\n");
    jseu_eval_cstr(a,
        "var o = {}; for (var i = 0; i < 200; i++) o['k' + i] = i;", NULL);
    jseu_eval_cstr(b,
        "var o = {}; for (var i = 0; i < 200; i++) o['k' + i] = i * 10;", NULL);
    show(a, "  A.o.k199", "o.k199");
    show(b, "  B.o.k199", "o.k199");
    show(a, "  A key count", "Object.keys(o).length");

    /*
     * -------------------------------------------------------- 4. strings
     *
     * Strings are interned per runtime, so the same literal is a different
     * HString in each. Equality is only ever asked within one runtime, which
     * is why that costs nothing.
     */
    printf("\nindependent string interning:\n");
    jseu_eval_cstr(a, "var s = 'shared-literal' + '';", NULL);
    jseu_eval_cstr(b, "var s = 'shared-literal' + '';", NULL);
    show(a, "  A.s === literal", "s === 'shared-literal'");
    show(b, "  B.s === literal", "s === 'shared-literal'");

    /*
     * -------------------------------------------------------- 5. handles
     *
     * A jse_value is an index into ONE runtime's registry, not a pointer to a
     * value, and the index is NOT tagged with which runtime it came from. Two
     * runtimes at the same allocation state hand out bit-identical handles, as
     * the first line below shows.
     *
     * So passing a handle to the wrong runtime's reader is NOT reliably caught.
     * It is caught only when that slot happens to be free or to carry a
     * different generation in the other runtime; when both registries are in
     * step, the read succeeds and quietly answers with the OTHER runtime's
     * value. Both outcomes are printed below, from the same pair of runtimes.
     *
     * Pairing a handle with its runtime is therefore the host's job, and the
     * engine will not check it for you. To move a value across, read it out on
     * one side and write it back on the other; no handle means anything in
     * both.
     */
    printf("\nhandles are per-runtime, and mixing them is not diagnosed:\n");

    /*
     * A fresh pair, so both registries start in step and issue the same
     * handle. `a` and `b` above have each allocated a different number of
     * handles by now, which would hide the collision behind a mismatched
     * generation tag and make this section look safer than it is.
     */
    {
        jse_runtime c = NULL, d = NULL;
        jse_value   vc = JSE_INVALID_VALUE, vd = JSE_INVALID_VALUE;

        if (jse_open(&c) != JSE_OK || jse_open(&d) != JSE_OK ||
            jseu_eval_cstr(c, "40 + 2", &vc) != JSE_OK ||
            jseu_eval_cstr(d, "7", &vd) != JSE_OK) {
            fprintf(stderr, "setting up the handle demo failed\n");
            jse_close(c);
            jse_close(d);
            goto fail;
        }

        printf("%-30s = %u vs %u (identical: %s)\n", "  C's handle vs D's handle",
               vc, vd, vc == vd ? "yes" : "no");

        /* Correct use: each handle read by the runtime that issued it. */
        n = -1.0;
        jse_get_number(c, vc, &n);
        printf("%-30s = %g\n", "  C's handle read by C", n);
        n = -1.0;
        jse_get_number(d, vd, &n);
        printf("%-30s = %g\n", "  D's handle read by D", n);

        /*
         * The silent case. Both registries are in step, so D resolves C's
         * handle and answers with its OWN value, 7, and returns JSE_OK.
         */
        n = -1.0;
        status = jse_get_number(d, vc, &n);
        printf("%-30s = %s, n=%g  <-- D's value, not C's\n",
               "  C's handle read by D", jseu_status_name(status), n);

        /*
         * The caught case, from the same pair. Freeing a handle in D moves its
         * generation counter on, so C's handle now names a slot D considers
         * stale and D rejects it. Identical mistake; only the registries'
         * relative state differs, which is why this is not a check you can
         * lean on.
         */
        jse_value_free(d, vd);
        n = -1.0;
        status = jse_get_number(d, vc, &n);
        printf("%-30s = %s, n=%g  <-- caught, only by luck\n",
               "  C's handle read by D again", jseu_status_name(status), n);

        jse_value_free(c, vc);
        jse_close(c);
        jse_close(d);
    }

    /*
     * Moving a value across is a read on one side and a write on the other.
     * `n` is read out of A through A's own reader; it is a plain C double by
     * then and belongs to neither runtime.
     */
    status = jseu_eval_cstr(a, "40 + 2", &from_a);
    if (status != JSE_OK) {
        fprintf(stderr, "eval in A failed: %s\n", jse_last_error(a));
        goto fail;
    }
    if (jse_get_number(a, from_a, &n) == JSE_OK) {
        char src[64];
        snprintf(src, sizeof(src), "var moved = %.17g;", n);
        jseu_eval_cstr(b, src, NULL);
        show(b, "  moved A->B via C", "moved");
    }

    /*
     * ------------------------------------------------------- 6. lifetime
     *
     * Closing one runtime invalidates only its own handles and leaves the
     * other entirely alone.
     */
    printf("\nclosing A leaves B alone:\n");
    jse_value_free(a, from_a);
    from_a = JSE_INVALID_VALUE;
    jse_close(a);
    a = NULL;
    show(b, "  B.tag after A closed", "tag + '/' + n + '/' + o.k199");

    jse_close(b);
    return EXIT_SUCCESS;

fail:
    if (from_a != JSE_INVALID_VALUE) {
        jse_value_free(a, from_a);
    }
    jse_close(a);
    jse_close(b);
    return EXIT_FAILURE;
}

File: bindings/zig/example/two_runtimes.zig
//! Two runtimes open at once: independent globals and objects, a host function
//! registered in both, and a value from one refused by the other.
//!
//! Each runtime has its own globals, objects, shapes and interned strings and
//! shares nothing with the other. A runtime must be driven from one thread at a
//! time -- the engine has no locking -- but two threads each driving their own
//! runtime share no state.

const std = @import("std");
const js = @import("jse");

/// Registered in both runtimes. `ctx.runtime()` tells the call which runtime it
/// is running in, so one function body can serve several.
fn label(ctx: js.Ctx, tag: *const []const u8) void {
    ctx.returnString(tag.*);
}

pub fn main(init: std.process.Init) !void {
    const gpa = init.arena.allocator();

    var out_buf: [4096]u8 = undefined;
    var out_file: std.Io.File.Writer = .init(.stdout(), init.io, &out_buf);
    const out = &out_file.interface;
    defer out.flush() catch {};

    // Two runtimes, open simultaneously.
    var a = try js.Runtime.init();
    defer a.deinit();
    var b = try js.Runtime.init();
    defer b.deinit();

    // --- independent globals ----------------------------------------------

    try a.exec("var tag = 'alpha-A'; var n = 111");
    try b.exec("var tag = 'beta-B';  var n = 222");

    var a_tag = try a.eval("tag + '/' + n");
    defer a_tag.deinit();
    var b_tag = try b.eval("tag + '/' + n");
    defer b_tag.deinit();
    try out.print("A globals = {s}\n", .{try a_tag.toString(gpa)});
    try out.print("B globals = {s}\n", .{try b_tag.toString(gpa)});

    // --- independent objects ----------------------------------------------
    //
    // The same property sequence in both, so the shape transitions interleave.

    try a.exec("var o = {}; for (let i = 0; i < 200; i++) o['k' + i] = i");
    try b.exec("var o = {}; for (let i = 0; i < 200; i++) o['k' + i] = i * 2");

    var a_k = try a.eval("o.k199");
    defer a_k.deinit();
    var b_k = try b.eval("o.k199");
    defer b_k.deinit();
    try out.print("A o.k199 = {d}, B o.k199 = {d}\n", .{
        try a_k.toNumber(),
        try b_k.toNumber(),
    });

    // --- one host function, two runtimes ----------------------------------

    const a_name: []const u8 = "from-A";
    const b_name: []const u8 = "from-B";
    try a.registerWith("label", label, &a_name, .{});
    try b.registerWith("label", label, &b_name, .{});

    var a_lbl = try a.eval("label()");
    defer a_lbl.deinit();
    var b_lbl = try b.eval("label()");
    defer b_lbl.deinit();
    try out.print("A label() = {s}, B label() = {s}\n", .{
        try a_lbl.toString(gpa),
        try b_lbl.toString(gpa),
    });

    // --- values do not cross runtimes -------------------------------------
    //
    // A handle is an index into ONE runtime's registry, and both runtimes
    // number their slots from the same base -- the handles below are equal as
    // integers. Offering a handle to the wrong runtime is a programming error:
    // it is caught only when that slot happens to be empty in the receiving
    // runtime, and otherwise reads out whatever unrelated value lives there.
    // Keep each value with its own runtime; do not rely on being told.

    var a_n = try a.eval("n");
    defer a_n.deinit();
    var b_n = try b.eval("n");
    defer b_n.deinit();
    try out.print("A handle {d} = {d}, B handle {d} = {d}\n", .{
        a_n.handle, try a_n.toNumber(),
        b_n.handle, try b_n.toNumber(),
    });

    // Same integer, different runtime, different value: nothing about the
    // handle itself distinguishes them.
    const smuggled = a_n.rebind(&b);
    if (smuggled.toNumber()) |v| {
        try out.print("A's handle read by B = {d} (B's own value, not A's)\n", .{v});
    } else |err| {
        try out.print("A's handle read by B = {s} (slot empty in B)\n", .{@errorName(err)});
    }

    // To move a value, read it out of one runtime and write it back into the
    // other. Numbers, strings and JSON all travel fine.
    var json = try a.eval("JSON.stringify({ from: tag, n })");
    defer json.deinit();
    const payload = try json.toString(gpa);

    var script: std.ArrayList(u8) = .empty;
    try script.appendSlice(gpa, "var imported = ");
    try script.appendSlice(gpa, payload);
    try script.appendSlice(gpa, "; imported.from + ' carried ' + imported.n");

    var carried = try b.eval(script.items);
    defer carried.deinit();
    try out.print("moved into B = {s}\n", .{try carried.toString(gpa)});

    // --- lifetimes are independent ----------------------------------------

    a.deinit();
    var after = try b.eval("tag + ' still works, n=' + n");
    defer after.deinit();
    try out.print("after closing A: {s}\n", .{try after.toString(gpa)});
}
